pkg/partitionprune/filter.go

```diff
@@ -0,0 +1,794 @@
+// Copyright 2022 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package partitionprune
+
+import (
+	"context"
+	"sort"
+
+	"github.com/matrixorigin/matrixone/pkg/sql/plan/function"
+
+	"github.com/matrixorigin/matrixone/pkg/container/batch"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/pb/partition"
+	"github.com/matrixorigin/matrixone/pkg/pb/plan"
+	"github.com/matrixorigin/matrixone/pkg/sql/colexec"
+	p "github.com/matrixorigin/matrixone/pkg/sql/plan"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+)
+
+func Filter(
+	proc *process.Process,
+	filters []*plan.Expr,
+	metadata partition.PartitionMetadata,
+) ([]int, error) {
+	if len(filters) == 0 {
+		res := make([]int, len(metadata.Partitions))
+		for i, pt := range metadata.Partitions {
+			res[i] = int(pt.Position)
+		}
+		return res, nil
+	}
+	switch metadata.Method {
+	case partition.PartitionMethod_Range:
+		return rangeFilter(proc, filters, metadata)
+	case partition.PartitionMethod_Hash,
+		partition.PartitionMethod_Key:
+		return hashFilter(proc, filters, metadata)
+	case partition.PartitionMethod_List:
+		return listFilter(proc, filters, metadata)
+	}
+	return nil, nil
+}
+
+func hashFilter(
+	proc *process.Process,
+	filters []*plan.Expr,
+	metadata partition.PartitionMetadata,
+) ([]int, error) {
+	colPosition := mustGetColPosition(metadata.Partitions[0].Expr)
+	tm := map[int]struct{}{}
+	for _, expr := range filters {
+		targets, ok, err := hashFilterExpr(proc, colPosition, expr, metadata)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			res := make([]int, len(metadata.Partitions))
+			for i, pt := range metadata.Partitions {
+				res[i] = int(pt.Position)
+			}
+			return res, nil
+		}
+		for _, target := range targets {
+			tm[target] = struct{}{}
+		}
+	}
+	res := make([]int, 0, len(tm))
+	for target := range tm {
+		res = append(res, target)
+	}
+	return res, nil
+}
+
+func hashFilterExpr(
+	proc *process.Process,
+	colPosition int32,
+	expr *plan.Expr,
+	metadata partition.PartitionMetadata,
+) ([]int, bool, error) {
+	exprs := make([]*plan.Expr, len(metadata.Partitions))
+	for i, pt := range metadata.Partitions {
+		exprs[i] = p.DeepCopyExpr(pt.Expr)
+	}
+	switch exprImpl := expr.Expr.(type) {
+	case *plan.Expr_F:
+		switch exprImpl.F.Func.ObjName {
+		case "or":
```

Comment:
What is the difference between "or" and "and" case?   Just how you combine them? 

This code should be simplified.

Reply:
For OR operator, recursively evaluate both left and right expressions and merge their results to get all matching partitions.
For AND operator, recursively evaluate both left and right expressions and find the intersection of their results to get matching partitions.